comm.go

```diff
@@ -44,6 +46,9 @@ func (p *PubSub) handleNewStream(s inet.Stream) {
 			return
 		}
 
+		measure := metrics.MIncomingMsgs.M((int64)(rpc.Size()))
+		stats.Record(context.Background(), measure)
```

Comment:
Please abstract both these calls behind a single `metrics` interface. I don't want to see this dance sprinkled everywhere in the code and it shouldn't detail on opencensus details whatsoever.

Reply:
Also, do we care about the size of messages? We care about the _number_ of messages a lot more.

Reply:
I wonder if it's worth abstracting what's already supposed to be an abstraction. It's already painful enough trying to use Prometheus through the keyhole that is OC. If OC is not a good abstraction, why are we using it?

Reply:
It doesn't matter, it's just metrics -- but I don't want to pollute the code with details of the metrics system.

Reply:
going to update what we're actually measuring here, so don't worry about that. imo, wrapping OC would be a lot of duplicated effort leading us straight back to go-metrics-interface, at which point all we've really done is create a new dependency. this means every measure would have to be accompanied by a bespoke function or we make a new interface

Reply:
Just turn the two functions into a single metrics call; I don't mind having duplication in the `metrics` package, I just don't want to see the god damn details of the metrics subsystem littered in the code!

Reply:
sorry about the tone, I didn't mean to sound so harsh. But I do feel strongly about it, the metrics details are just that: implementation details and I shouldn't have to write two lines of that just to increment a counter.

Reply:
no worries and yeah, it does feel a bit ugly. i'll put up some proposals

Reply:
OC *is* the abstraction layer and we've picked it because it's an open standard. Introducing an abstraction layer over an abstraction layer not only has runtime cost, but it's redundant.

Reply:
Let's be [DRY](https://en.wikipedia.org/wiki/Don't_repeat_yourself) and encapsulate common OC rituals behind a package (github.com/libp2p/go-libp2p-core/metrics) -- but let's not create _an abstraction_ but a helper toolbox.

Reply:
I would argue that it's a very poor interface. I don't want to care about the details of metrics in the code, so hide it behind the dedicated module that handles them. Whether that's `core/metrics` or the `metrics` package which we have in here, it's immaterial.

Reply:
Also, the runtime cost here is negligible. I am not asking for a new abstraction layer, I want a better interface that doesn't pollute the code with _crap_ :p

Reply:
Thoughts:

1. We don't want to _pollute_ the code with **repetitive** ceremonies. But if the OC call is a one-liner, we are not going to put it behind an indirection. There's no reason to hide OC -- but there is a reason to be DRY.

2. Stats bookkeeping will be done _everywhere, all the time_. The runtime cost will not be negligible and we should check all calls the library we cook up are inlined (-gcflags="-m") and that we're _not allocating_ beyond what OC needs.

Reply:
Note that I am not asking for an abstraction here, I am asking to make named methods that hide the details of the OC ugliness.
Something like `metrics.inrementMessageCounter()` that hides the two lines is good enough for me.

Reply:
But if we have a nicer abstraction, that would obviously be acceptable.

Bottom line: I don't want to have to care about `M` or contexts here.